tools/netpipe-2.4/netpipe.1
.\" -*- nroff -*-
.\"
.\" NetPIPE -- Network Protocol Independent Performance Evaluator.
.\" Copyright 1997, 1998 Iowa State University Research Foundation, Inc.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation.  You should have received a copy of the
.\" GNU General Public License along with this program; if not, write to the
.\" Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.\"
.\" netpipe.1
.\" Created: Mon Jun 15 1998 by Guy Helmer
.\"
.TH netpipe 1 "June 15, 1998" "NetPIPE" "netpipe"

.SH NAME
NetPIPE \- network protocol independent performance evaluator

.SH SYNOPSIS
.B NPtcp
[\c
.BI \-A \ buffer_alignment\fR\c
]
[\c
.BR \-a \c
]
[\c
.BI \-b \ TCP_buffer_size\fR\c
]
[\c
.BI \-h \ host_name\fR\c
]
[\c
.BI \-i \ increment\fR\c
]
[\c
.BI \-l \ starting_msg_size\fR\c
]
[\c
.BI \-O \ buffer_offset\fR\c
]
[\c
.BI \-o \ output_filename\fR\c
]
[\c
.BR \-P \c
]
[\c
.BI \-p \ port\fR\c
]
[\c
.BR \-r \c
]
[\c
.BR \-s \c
]
[\c
.BR \-t \c
]
[\c
.BI \-u \ ending_msg_size\fR\c
]

.PP

.B NPmpi
[\c
.BI \-A \ buffer_alignment\fR\c
]
[\c
.BR \-a \c
]
[\c
.BI \-i \ increment\fR\c
]
[\c
.BI \-l \ starting_msg_size\fR\c
]
[\c
.BI \-O \ buffer_offset\fR\c
]
[\c
.BI \-o \ output_filename\fR\c
]
[\c
.BR \-P \c
]
[\c
.BR \-s \c
]
[\c
.BI \-u \ ending_msg_size\fR\c
]

.PP

.B NPpvm
[\c
.BI \-A \ buffer_alignment\fR\c
]
[\c
.BR \-a \c
]
[\c
.BI \-i \ increment\fR\c
]
[\c
.BI \-l \ starting_msg_size\fR\c
]
[\c
.BI \-O \ buffer_offset\fR\c
]
[\c
.BI \-o \ output_filename\fR\c
]
[\c
.BR \-P \c
]
[\c
.BR \-r \c
]
[\c
.BR \-s \c
]
[\c
.BR \-t \c
]
[\c
.BI \-u \ ending_msg_size\fR\c
]

.SH DESCRIPTION
.PP
.B NetPIPE
is a protocol independent performance tool that encapsulates
the best of ttcp and netperf and visually represents the network
performance under a variety of conditions. By taking the end-to-end
application view of a network,
.B NetPIPE
clearly shows the overhead
associated with different protocol layers.
.B NetPIPE
answers such questions as:
.RS
How soon will a given data block of size k arrive at its destination?
.PP
Which network and protocol will transmit size k blocks the fastest?
.PP
What is a given network's effective maximum throughput and saturation
level?
.PP
Does there exist a block size k for which the throughput is maximized?
.PP
How much communication overhead is due to the network communication
protocol layer(s)?
.PP
How quickly will a small (< 1 kbyte) control message arrive, and which
network and protocol are best for this purpose?
.RE
.PP
.B NetPIPE
is provided with interfaces for TCP, MPI, and PVM, but TCP is the most
commonly used interface for general network testing purposes.  It
should be easy to write new interfaces for other reliable protocols based
on the examples provided by the TCP, MPI and PVM interfaces.
.SH TESTING TCP
.PP
Typical use for TCP involves running the TCP NetPIPE receiver on one
system with the command
.PP
.Ex
NPtcp \-r
.Ee
.PP
and running the TCP NetPIPE transmitter on another system with the
command
.PP
.Ex
NPtcp \-h receiver_hostname \-o output_filename \-P \-t
.Ee
.PP
If any options are used that modify the test protocol, including \-i,
\-l, \-p, \-s, and \-u, those parameters
.B must
be used on both the transmitter and the receiver, or the test
will not run properly.
.SH TESTING PVM
.PP
Typical use for PVM first requires starting PVM with the command
.PP
.Ex
pvm
.Ee
.PP
and adding a second machine with the PVM command
.PP
.Ex
add othermachine
.Ee
.PP
(then exit the PVM command line interface).  Then run the PVM NetPIPE
receiver on one system with the command
.PP
.Ex
NPpvm \-r
.Ee
.PP
and run the TCP NetPIPE transmitter on the other system with the
command
.PP
.Ex
NPpvm \-t \-o output_filename \-P
.Ee
.PP
If any options are used that modify the test protocol, including \-i,
\-l, \-p, \-s, and \-u, those parameters
.B must
be used on both the transmitter and the receiver, or the test
will not run properly.
.SH TESTING MPI
.PP
Use of the MPI interface for NetPIPE depends on the MPI implementation
used.  For the Argonne MPICH implementation using the p4 device (for a
cluster of individual systems interconnected using TCP/IP), create a
file that contains the hostnames of the two systems you want to
include in the test, with one hostname on each line of the file
(assume the file is named "machines.p4").  Then, use the command
.PP
.Ex
mpirun \-machinefile machines.p4 \-np 2 NPmpi \-o output_filename \-P
.Ee
.PP
to start the test.  MPICH will start an NPmpi process on each of the
two selected machines and the test will begin.
.SH TESTING METHODOLOGY
.PP
.B NetPIPE
tests network performance by sending a number of messages at each
block size, starting from the lower bound on message size.
.B NetPIPE
increments the message size until the upper bound on message size is
reached or the time to transmit a block exceeds one second, which ever
occurs first.
.PP
.B NetPIPE\c
\'s output file may be graphed with a program such as
.B gnuplot(1)
to view the results of the test.
.B NetPIPE\c
\'s
output file contains five columns: time to transfer the block, bits
per second, bits in block, bytes in block, and variance.  These
columns may be graphed to represent and compare the network's
performance.  For example, the
.B network signature graph
can be created by graphing time versus bits per second.  Sample
.B gnuplot(1)
commands for such a graph would be
.PP
.Ex
set logscale x
.Ee
.PP
.Ex
plot "NetPIPE.out" using 1:2
.Ee
.PP
The more traditional
.B throughput versus block size
graph can be created by graphing bytes versus bits per second.
Sample
.B gnuplot(1)
commands for such a graph would be
.PP
.Ex
set logscale x
.Ee
.PP
.Ex
plot "NetPIPE.out" using 4:2
.Ee

.ne 5
.SH OPTIONS
.TP
.B \-A \ \fIalignment\fR
Align buffers to the given boundary.  For example, a value of 4 would
align buffers to 4-byte (word) boundaries.
.ne 3
.TP
.B \-a
Specify asynchronous receive (a.k.a. preposted receive), if the
underlying protocol supports it.
.ne 3
.TP
.BI \-b \ \fIbuffer_size\fR
[TCP only] Set send and receive TCP buffer sizes.
.ne 3
.TP
.BI \-h \ \fIhostname\fR
[TCP transmitter only] Specify name of host to which to connect.
.ne 3
.TP
.BI \-i \ \fIincrement\fR
Specify increment step size (default is an exponentially increasing
increment).
.ne 3
.TP
.BI \-l \ \fIstart_msg_size\fR
Specify the starting message size.  The test will start with messages
of this size and increment, either exponentially or with an increment
specified by the
.B \-i
flag, until a block requires more than one second to transmit or the
ending message size specified by the
.B \-u
flag is reached, which ever occurs first.
.ne 3
.TP
.BI \-O \ \fIbuffer_offset\fR
Specify offset of buffers from alignment.  For example, specifying an
alignment of 4 (with \-A) and an offset of 1 would align buffers to
the first byte after a word boundary.
.ne 3
.TP
.BI \-o \ \fIoutput_filename\fR
Specify output filename.  By default, the output filename is
.IR NetPIPE.out .
.ne 3
.TP
.B \-P
Print results on screen during execution of the test.  By default,
NetPIPE is silent during execution of the test.
.ne 3
.TP
.BI \-p \ \fIport_number\fR
[TCP only] Specify TCP port number to which to connect (for the
transmitter) or the port on which to listen for connections (for the
receiver).
.ne 3
.TP
.B \-r
[TCP only] This process is a TCP receiver.
.ne 3
.TP
.B \-s
Set streaming mode: data is only transmitted in one direction.  By
default, the transmitter measures the time taken as each data block is
sent from the transmitter to the receiver and back, then divides the
round-trip time by two to obtain the time taken by the message to
travel in each direction.  In streaming mode, the receiver measures
the time required to receive the message and sends the measured time
back to the transmitter for posting to the output file.
.ne 3
.TP
.B \-t
[TCP only] This process is a TCP transmitter.
.ne 3
.TP
.BI \-u \ \fIending_msg_size\fR
Specify the ending message size.  By default, the test will end when
the time to transmit a block exceeds one second.  If
.B \-u
is specified, the test will end when either the test time exceeds one
second or the ending message size is reached, which ever occurs first.

.ne 3
.SH FILES
.TP
.I NetPIPE.out
Default output file for
.BR NetPIPE .
Overridden by the
.B \-o
option.

.SH AUTHOR
.PP
Quinn Snell <snell@cs.byu.edu>, Guy Helmer <ghelmer@scl.ameslab.gov>,
and others.
.PP
Clark Dorman <dorman@s3i.com> contributed the PVM interface.
.PP
Information about
.B NetPIPE
can be found on the World Wide Web at
http://www.scl.ameslab.gov/netpipe/.

.SH BUGS
By nature,
.B NetPIPE
will use as much of the network bandwidth as possible.  Other users of
the network may notice the effect.


ChangeLog
LTP-20100831

commit dd59d49d75598d163f1b4599067a9f0dcb70e6c4
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 23:09:49 2010 -0700

    The cacheflush system call does not return EINVAL in the latest Linux
    kernel. see the link:
    A related patch about the cacheflush function:
    http://lkml.org/lkml/2009/4/9/203
    But Ralf had refused this patch:
    http://www.spinics.net/lists/linux-man/msg00906.html
    cacheflush01 testcase checks if the return value is EINVAL when the
    cache argument is not one of ICACHE, BCACHE or DCACHE. So it will fail
    in all boards.
    
    In this modification, checking return value SUCCESS will be added,
    instead of checking EINVAL
    
    Signed-off-by: Feng Gao <feng.gao@windriver.com>

:100644 100644 cce0b70... d0a4b1f... M	testcases/kernel/syscalls/cacheflush/cacheflush01.c

commit ebea45b92a1e30e4480f25291ef4c1ef2eeaa4bf
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Tue Aug 31 17:14:21 2010 -0700

    Get rid of unnecessary style divergence.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 0ebeca9... b08a2d9... M	testcases/commands/at/at_allow01
:100755 100755 6bca177... 9f9f433... M	testcases/commands/at/at_deny01

commit 08640ca184c1fc1e24e966c3d806a722f3ca2154
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:53:18 2010 -0700

    Add more assertions.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:000000 100644 0000000... 2c590b1... A	testcases/open_posix_testsuite/conformance/interfaces/access/assertions.xml

commit 3cdde2f86af90b3fd4e2b1468a12adfd2a9523f0
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 20:52:52 2010 -0700

    Note which timer failed in output.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 ac043b0... 8e149ea... M	testcases/open_posix_testsuite/conformance/behavior/timers/1-1.c

commit 207f47c4c9b4783c1a18be32431f5a1b319b9e5a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 19:34:34 2010 -0700

    Update some assertions for raise(2) to SUSv4 reqs.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 38d9542... aa47b7c... M	testcases/open_posix_testsuite/conformance/interfaces/raise/assertions.xml

commit 9c6d2b4cf4f7c166813ddee280ab2bf128eb521c
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:36:51 2010 -0700

    Fix bad syntax that I thought I hadn't committed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 f731d6c... 0ebeca9... M	testcases/commands/at/at_allow01

commit 414747ced26d943a22a3f767deb7f5b4658adb6e
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 03:32:18 2010 -0700

    Forgot the sleeps.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 9b10030... 6d5eb7f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 a3be617... 60e5241... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c
:100644 100644 34e3afd... abc3279... M	testcases/open_posix_testsuite/conformance/interfaces/pthread_attr_setstack/2-1.c
:100755 100755 01984c5... 536d407... M	testcases/open_posix_testsuite/scripts/generate-makefiles.sh

commit 3222fbecbfb2924a6b901a8e2048c70e1a993018
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:22:07 2010 -0700

    Remove bogus comment.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 c7d81b5... 72a1fcd... M	testcases/open_posix_testsuite/conformance/behavior/timers/2-1.c

commit 7187f0921382bf375cd111b8a480a74ce5786945
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:18:23 2010 -0700

    Repeat robustness work done to 1-1 for 1-2 and 1-3.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 a9d104c... 9b10030... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-2.c
:100644 100644 af23ebe... a3be617... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-3.c

commit 6c2750a55eaa30fe708c78087688c0d468bfb807
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sun Aug 29 01:12:07 2010 -0700

    Make test more robust.
    
    1. Catch fork() == -1 errors.
    2. Sleep so wait(2) should always succeed.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100644 100644 1efd520... f46ee8f... M	testcases/open_posix_testsuite/conformance/behavior/WIFEXITED/1-1.c

commit 873f6e5b28a6a23367a7117aad436378e56ad09a
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:42:53 2010 -0700

    Fix ltp-aiodio.sh script so it execs aio-stress.
    
    Singed-off-by: Jacky Malcles <jacky.malcles@bull.net>

:100755 100755 ed2bce4... ab54c54... M	testscripts/ltp-aiodio.sh

commit 4fd92183833eed128bb6420ece7a6dd45dcd8ec5
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:40:51 2010 -0700

    I've prepared a patch for at_allow01 that essentially duplicates the
    fixes that were checked in for at_deny01. This fixes the script to
    actually run the test when the at utility in stead of returning TCONF,
    correcting test environment, and let the script run normally without
    changing the password. See commit
    a589d12b0b49287d10f43b1a1fb72950819588eb for more details.
    
    Signed-off-by: Henry Yei <hyei@mvista.com>

:100755 100755 cf5b97c... f731d6c... M	testcases/commands/at/at_allow01

commit 217d9b778d275bcbaf5a9fa9e1a5ac2716521cad
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Sat Aug 28 01:39:12 2010 -0700

    Chain cmds together to avoid intermediate errors.
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>

:100755 100755 b94bd8f... 6bca177... M	testcases/commands/at/at_deny01

commit 9eeea7939bd73b295ae376dc045e5fbb984d0f38
Author: Garrett Cooper <yanegomi@gmail.com>
Date:   Wed Aug 25 20:40:23 2010 -0700

    Fix error message (was hardcoded to SIGALARM).
    
    Signed-off-by: Garrett Cooper <yanegomi@gmail.com>


tools/netpipe-2.4-ipv6/netpipe.c
/*****************************************************************************/
/* "NetPIPE" -- Network Protocol Independent Performance Evaluator.          */
/* Copyright 1997, 1998 Iowa State University Research Foundation, Inc.      */
/*                                                                           */
/* This program is free software; you can redistribute it and/or modify      */
/* it under the terms of the GNU General Public License as published by      */
/* the Free Software Foundation.  You should have received a copy of the     */
/* GNU General Public License along with this program; if not, write to the  */
/* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.   */
/*                                                                           */
/* Files needed for use:                                                     */
/*     * netpipe.c          ---- Driver source                               */
/*     * netpipe.h          ---- General include file                        */
/*     * TCP.c              ---- TCP calls source                            */
/*     * TCP.h              ---- Include file for TCP calls and data structs */
/*     * MPI.c              ---- MPI calls source                            */
/*     * MPI.h              ---- Include file for MPI calls and data structs */
/*     * PVM.c              ---- PVM calls source                            */
/*     * PVM.h              ---- Include file for PVM calls and data structs */
/* 2002/03/18 --- Modified to specify server interfaces - Robbie Williamson  */
/*		  (robbiew@us.ibm.com)					     */
/*****************************************************************************/
#include "netpipe.h"

extern char *optarg;


/* Return the current time in seconds, using a double precision number.      */
double
When()
{
    struct timeval tp;
    gettimeofday(&tp, NULL);
    return ((double) tp.tv_sec + (double) tp.tv_usec * 1e-6);
}

int PrintUsage()
{
	printf("\n NETPIPE USAGE \n\n");
	printf("A: specify buffers alignment e.g.: <-A 1024>\n");
	printf("a: asynchronous receive (a.k.a. preposted receive)\n");
#if defined(TCP)
	printf("b: specify send and receive buffer sizes e.g. <-b 32768>\n");
	printf("h: specify hostname <-h host>\n");
	printf("H: specify server hostname <-H HOST> e.g. Multiple NICs\n");
#endif
	printf("i: specify increment step size e.g. <-i 64>\n");
	printf("l: lower bound start value e.g. <-i 1>\n");
	printf("O: specify buffer offset e.g. <-O 127>\n");
	printf("o: specify output filename <-o fn>\n");
	printf("P: print on screen\n");
#if defined(TCP)
	printf("p: specify port e.g. <-p 5150>\n");
#endif
	printf("r: receiver\n");
	printf("s: stream option\n");
	printf("t: transmitter\n");
	printf("u: upper bound stop value e.g. <-u 1048576>\n");
	printf("\n");
	exit(-12);
}

int main(int argc, char *argv[])
{
    FILE        *out;           /* Output data file                          */
    char        s[255];         /* Generic string                            */
    char        *memtmp;
    char        *memtmp1;

    int         c,              /* option index                              */
                i, j, n, nq,    /* Loop indices                              */
		asyncReceive=0, /* Pre-post a receive buffer?                */
                bufoffset=0,    /* Align buffer to this                      */
                bufalign=16*1024,/* Boundary to align buffer to              */
		errFlag,        /* Error occurred in inner testing loop      */
                nrepeat,        /* Number of time to do the transmission     */
                len,            /* Number of bytes to be transmitted         */
                inc=0,          /* Increment value                           */
                trans=-1,       /* Transmitter flag. 1 if transmitting.      */
                server=0,       /* Server flag. 1 if specifying server.      */
                detailflag=0,   /* Set to examine the signature curve detail */
                bufszflag=0,    /* Set to change the TCP socket buffer size  */
                pert,           /* Perturbation value                        */
                start=1,        /* Starting value for signature curve        */
                end=MAXINT,     /* Ending value for signature curve          */
                streamopt=0,    /* Streaming mode flag                       */
                printopt=0;     /* Debug print statements flag               */

    ArgStruct   args;           /* Argumentsfor all the calls                */

    double      t, t0, t1, t2,  /* Time variables                            */
                tlast,          /* Time for the last transmission            */
                latency;        /* Network message latency                   */

    Data        bwdata[NSAMP];  /* Bandwidth curve data                      */

    short       port=DEFPORT;   /* Port number for connection                */
#ifdef HAVE_GETRUSAGE
    struct rusage prev_rusage, curr_rusage; /* Resource usage                */
    double	user_time, sys_time;	/* User & system time used                   */
    double	best_user_time, best_sys_time; /* Total user & system time used     */
    double      ut1, ut2, st1, st2; /* User & system time ctrs for variance  */
    double	ut_var, st_var;	/* Variance in user & system time            */
#endif

#ifdef MPI
    MPI_Init(&argc, &argv);
#endif

    strcpy(s, "NetPIPE.out");
#ifndef MPI
    if (argc < 2)
     PrintUsage();
#endif
    /* Parse the arguments. See Usage for description */
    while ((c = getopt(argc, argv, "PstrhH:p:o:A:O:l:u:i:b:a")) != -1)
    {
        switch(c)
        {
            case 'o': strcpy(s,optarg);
                      break;

            case 't': trans = 1;
                      break;

            case 'r': trans = 0;
                      break;

            case 's': streamopt = 1;
                      break;

            case 'l': /*detailflag = 1;*/
                      start = atoi(optarg);
                      if (start < 1)
                      {
                        fprintf(stderr,"Need a starting value >= 1\n");
                        exit(743);
                      }
                      break;

            case 'u': /*detailflag = 1;*/
                      end = atoi(optarg);
                      break;

            case 'i': detailflag = 1;
                      inc = atoi(optarg);
                      break;

            case 'b': bufszflag = 1;
#ifdef TCP
                      args.prot.rcvbufsz=atoi(optarg);
                      args.prot.sndbufsz=args.prot.rcvbufsz;
#endif
                      break;

            case 'P': printopt = 1;
                      break;

            case 'A': bufalign = atoi(optarg);
                      break;

            case 'O': bufoffset = atoi(optarg);
                      break;

            case 'p': port = atoi(optarg);
                      break;

            case 'h': if (trans == 1)
                      {
                          args.host = (char *)malloc(strlen(optarg)+1);
                          strcpy(args.host, optarg);
			  printf("host is %s\n",args.host);
                      }
		      else
		      {
			  fprintf(stderr, "Error: -t must be specified before -h\n");
			  exit(-11);
		      }
                      break;

            case 'H': if (trans == 0)
                      {
                          args.server_host = (char *)malloc(strlen(optarg)+1);
                          strcpy(args.server_host, optarg);
			  printf("server is %s\n",args.server_host);
			  server = 1;
                      }
		      else
		      {
			  fprintf(stderr, "Error: -r must be specified before -H\n");
			  exit(-11);
		      }
                      break;

	    case 'a': asyncReceive = 1;
		      break;

            default:  PrintUsage();
                      exit(-12);
        }
    }
    if (start > end)
    {
        fprintf(stderr, "Start MUST be LESS than end\n");
        exit(420132);
    }
#if defined(TCP) || defined(PVM)
    /*
      It should be explicitly specified whether this is the transmitter
      or the receiver.
    */
    if (trans < 0)
    {
	fprintf(stderr, "Error: either -t or -r must be specified\n");
	exit(-11);
    }
#endif

    args.nbuff = TRIALS;
    args.tr = trans;
    args.sr = server;
    args.port = port;

#if defined(TCP)
    if (!bufszflag)
    {
        args.prot.sndbufsz = 0;
        args.prot.rcvbufsz = 0;
    }
    else
        fprintf(stderr,"Send and Recv Buffers are %d bytes\n",
                                                   args.prot.sndbufsz);
#endif

    Setup(&args);
    Establish(&args);

    if (args.tr)
    {
        if ((out = fopen(s, "w")) == NULL)
        {
            fprintf(stderr,"Can't open %s for output\n", s);
            exit(1);
        }
    }
    else
        out = stdout;

    args.bufflen = 1;
    args.buff = (char *)malloc(args.bufflen);
    args.buff1 = (char *)malloc(args.bufflen);
    if (asyncReceive)
	PrepareToReceive(&args);
    Sync(&args);
    t0 = When();
    t0 = When();
    t0 = When();
#ifdef HAVE_GETRUSAGE
    getrusage(RUSAGE_SELF, &prev_rusage);
#endif
    t0 = When();
    for (i = 0; i < LATENCYREPS; i++)
    {
        if (args.tr)
        {
            SendData(&args);
            RecvData(&args);
	    if (asyncReceive && (i < LATENCYREPS - 1))
	    {
		PrepareToReceive(&args);
	    }
        }
        else
        {
            RecvData(&args);
	    if (asyncReceive && (i < LATENCYREPS - 1))
	    {
		PrepareToReceive(&args);
	    }
            SendData(&args);
        }
    }
    latency = (When() - t0)/(2 * LATENCYREPS);
#ifdef HAVE_GETRUSAGE
    getrusage(RUSAGE_SELF, &curr_rusage);
#endif
    free(args.buff);
    free(args.buff1);


    if (args.tr)
    {
        SendTime(&args, &latency);
    }
    else
    {
        RecvTime(&args, &latency);
    }
    if (args.tr && printopt)
    {
        fprintf(stderr,"Latency: %.7f\n", latency);
        fprintf(stderr,"Now starting main loop\n");
    }
    tlast = latency;
    if (inc == 0)
    {
	/* Set a starting value for the message size increment. */
	inc = (start > 1) ? start / 2 : 1;
    }

    /* Main loop of benchmark */
    for (nq = n = 0, len = start, errFlag = 0;
         n < NSAMP - 3 && tlast < STOPTM && len <= end && !errFlag;
         len = len + inc, nq++ )
    {
        if (nq > 2 && !detailflag)
	  {
	    /*
	      This has the effect of exponentially increasing the block
	      size.  If detailflag is false, then the block size is
	      linearly increased (the increment is not adjusted).
	    */
            inc = ((nq % 2))? inc + inc: inc;
	  }

        /* This is a perturbation loop to test nearby values */
        for (pert = (!detailflag && inc > PERT+1)? -PERT: 0;
             pert <= PERT;
             n++, pert += (!detailflag && inc > PERT+1)? PERT: PERT+1)
        {

            /* Calculate how many times to repeat the experiment. */
            if (args.tr)
            {
                nrepeat = MAX((RUNTM / ((double)args.bufflen /
                                 (args.bufflen - inc + 1.0) * tlast)), TRIALS);
                SendRepeat(&args, nrepeat);
            }
            else
            {
                RecvRepeat(&args, &nrepeat);
            }

            /* Allocate the buffer */
            args.bufflen = len + pert;
            if ((args.buff=(char *)malloc(args.bufflen+bufalign))==(char *)NULL)
            {
                fprintf(stderr,"Couldn't allocate memory\n");
		errFlag = -1;
                break;
            }
            if ((args.buff1=(char *)malloc(args.bufflen+bufalign))==(char *)NULL)
            {
                fprintf(stderr,"Couldn't allocate memory\n");
		errFlag = -1;
                break;
            }
            /*
	      Possibly align the data buffer: make memtmp and memtmp1
	      point to the original blocks (so they can be freed later),
	      then adjust args.buff and args.buff1 if the user requested it.
	    */
            memtmp = args.buff;
            memtmp1 = args.buff1;
            if (bufalign != 0)
                args.buff +=(bufalign -
                        ((intptr_t)args.buff % bufalign) + bufoffset) % bufalign;

            if (bufalign != 0)
                args.buff1 +=(bufalign -
                        ((intptr_t)args.buff1 % bufalign) + bufoffset) % bufalign;


            if (args.tr && printopt)
                fprintf(stderr,"%3d: %9d bytes %4d times --> ",
                                 n,args.bufflen,nrepeat);

            /* Finally, we get to transmit or receive and time */
            if (args.tr)
            {
		/*
		   This is the transmitter: send the block TRIALS times, and
		   if we are not streaming, expect the receiver to return each
		   block.
		*/
                bwdata[n].t = LONGTIME;
                t2 = t1 = 0;
#ifdef HAVE_GETRUSAGE
		ut1 = ut2 = st1 = st2 = 0.0;
		best_user_time = best_sys_time = LONGTIME;
#endif
                for (i = 0; i < TRIALS; i++)
                {
                    Sync(&args);
#ifdef HAVE_GETRUSAGE
		    getrusage(RUSAGE_SELF, &prev_rusage);
#endif
                    t0 = When();
                    for (j = 0; j < nrepeat; j++)
                    {
			if (asyncReceive && !streamopt)
			{
			    PrepareToReceive(&args);
			}
                        SendData(&args);
                        if (!streamopt)
			{
                            RecvData(&args);
			}
                    }
                    t = (When() - t0)/((1 + !streamopt) * nrepeat);
#ifdef HAVE_GETRUSAGE
		    getrusage(RUSAGE_SELF, &curr_rusage);
		    user_time = ((curr_rusage.ru_utime.tv_sec -
			      prev_rusage.ru_utime.tv_sec) + (double)
			     (curr_rusage.ru_utime.tv_usec -
			      prev_rusage.ru_utime.tv_usec) * 1.0E-6) /
		      ((1 + !streamopt) * nrepeat);
		    sys_time = ((curr_rusage.ru_stime.tv_sec -
			      prev_rusage.ru_stime.tv_sec) + (double)
			     (curr_rusage.ru_stime.tv_usec -
			      prev_rusage.ru_stime.tv_usec) * 1.0E-6) /
		      ((1 + !streamopt) * nrepeat);
		    ut2 += user_time * user_time;
		    st2 += sys_time * sys_time;
		    ut1 += user_time;
		    st1 += sys_time;
		    if ((user_time + sys_time) < (best_user_time + best_sys_time))
		    {
			best_user_time = user_time;
			best_sys_time = sys_time;
		    }
#endif

                    if (!streamopt)
                    {
                        t2 += t*t;
                        t1 += t;
                        bwdata[n].t = MIN(bwdata[n].t, t);
                    }
                }
                if (!streamopt)
                    SendTime(&args, &bwdata[n].t);
                else
                    RecvTime(&args, &bwdata[n].t);

                if (!streamopt)
                    bwdata[n].variance = t2/TRIALS - t1/TRIALS * t1/TRIALS;

#ifdef HAVE_GETRUSAGE
		ut_var = ut2/TRIALS - (ut1/TRIALS) * (ut1/TRIALS);
		st_var = st2/TRIALS - (st1/TRIALS) * (st1/TRIALS);
#endif

            }
            else
            {
		/*
		   This is the receiver: receive the block TRIALS times, and
		   if we are not streaming, send the block back to the
		   sender.
		*/
                bwdata[n].t = LONGTIME;
                t2 = t1 = 0;
                for (i = 0; i < TRIALS; i++)
                {
		    if (asyncReceive)
		    {
		    	PrepareToReceive(&args);
		    }
                    Sync(&args);
                    t0 = When();
                    for (j = 0; j < nrepeat; j++)
                    {
                        RecvData(&args);
			if (asyncReceive && (j < nrepeat - 1))
			{
			    PrepareToReceive(&args);
			}
                        if (!streamopt)
                            SendData(&args);
                    }
                    t = (When() - t0)/((1 + !streamopt) * nrepeat);

                    if (streamopt)
                    {
                        t2 += t*t;
                        t1 += t;
                        bwdata[n].t = MIN(bwdata[n].t, t);
                    }
                }
                if (streamopt)
                    SendTime(&args, &bwdata[n].t);
                else
                    RecvTime(&args, &bwdata[n].t);

                if (streamopt)
                    bwdata[n].variance = t2/TRIALS - t1/TRIALS * t1/TRIALS;

            }
            tlast = bwdata[n].t;
            bwdata[n].bits = args.bufflen * CHARSIZE;
            bwdata[n].bps = bwdata[n].bits / (bwdata[n].t * 1024 * 1024);
            bwdata[n].repeat = nrepeat;

            if (args.tr)
	    {
		fprintf(out, "%.7f %.7f %d %d %.7f", bwdata[n].t, bwdata[n].bps,
			bwdata[n].bits, bwdata[n].bits / 8,
			bwdata[n].variance);
#ifdef HAVE_GETRUSAGE
		fprintf(out, " %.7f %.7f %.7f %.7f", ut1 / (double) TRIALS,
			st1 / (double) TRIALS, ut_var, st_var);
#endif
		fprintf(out, "\n");
	    }
            fflush(out);

            free(memtmp);
            free(memtmp1);

            if (args.tr && printopt)
	    {
        	fprintf(stderr," %6.3f Mbps in %.7f sec", bwdata[n].bps,
			tlast);
#ifdef HAVE_GETRUSAGE
		fprintf(stderr, ", avg utime=%.7f avg stime=%.7f, ",
			ut1 / (double) TRIALS,
			st1 / (double) TRIALS);
		fprintf(stderr, "min utime=%.7f stime=%.7f, ", best_user_time,
			best_sys_time);
		fprintf(stderr, "utime var=%.7f stime var=%.7f", ut_var, st_var);
#endif
		fprintf(stderr, "\n");
	    }
        } /* End of perturbation loop */

    } /* End of main loop  */

    if (args.tr)
       fclose(out);

    CleanUp(&args);
    return(0);
}
